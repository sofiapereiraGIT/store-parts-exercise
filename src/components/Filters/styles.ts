import styled from "styled-components";
import { TextField } from "@material-ui/core";
import { Autocomplete, IconButton } from "@mui/material";

export const Filters = {
  FiltersWrapper: styled.div`
    display: flex;
    flex-wrap: wrap;
    > div {
      flex: 1 33%;
      margin: 0.5em 1em 0.5em 0;
      min-width: 150px;
    }
  `,
  TextField: styled(TextField)`
    && {
      * {
        font-family: inherit;
        font-size: 1em;
      }
      .MuiOutlinedInput-input {
        padding: 1em 0.8em;
      }
      .MuiInputLabel-outlined {
        transform: translate(0.8em, 1em) scale(1);
      }
      .MuiInputLabel-outlined.MuiInputLabel-shrink {
        transform: translate(14px, -6px) scale(0.8);
      }
    }
  `,
  Dropdown: styled(Autocomplete)`
    && {
      .MuiAutocomplete-input {
        padding: min(0.4em, 6px) !important;
      }
      .MuiButtonBase-root {
        margin-top: 0.3em;
      }
    }
  `,
  IconButton: styled(IconButton)`
    && {
      padding: 0.8em;
    }
  `,
};
