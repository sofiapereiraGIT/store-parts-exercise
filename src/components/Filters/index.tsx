import ArrowUpwardIcon from "@mui/icons-material/ArrowUpward";
import ArrowDownwardIcon from "@mui/icons-material/ArrowDownward";
import { Filters as S } from "./styles";

export default function Filters(props: {
  searchFilter: string;
  setSearchFilter: (newSearchFilter: string) => void;
  loadingTypeOpt: boolean;
  typeOptions: string[];
  typeFilter: string;
  setTypeFilter: (newTypeFilter: string) => void;
  descOrder: boolean;
  onChangeDescOrder: () => void;
}) {
  return (
    <S.FiltersWrapper>
      <S.TextField
        id="search"
        inputProps={{ "data-testid": "searchField" }}
        type="search"
        variant="outlined"
        label="Search"
        value={props.searchFilter}
        onChange={(e) => props.setSearchFilter(e.target.value)}
      />
      <S.Dropdown
        id="autocomplete"
        options={props.typeOptions}
        onChange={(e, value) =>
          props.setTypeFilter(value === null ? "" : (value as string))
        }
        renderInput={(params) => (
          <S.TextField
            variant="outlined"
            label="Type"
            placeholder={props.loadingTypeOpt ? "loading..." : ""}
            value={props.typeFilter}
            {...params}
          />
        )}
      />
      <S.IconButton
        data-testid="descOrderIcon"
        onClick={props.onChangeDescOrder}
      >
        {props.descOrder ? (
          <ArrowUpwardIcon data-testid="arrowUpwardIcon" />
        ) : (
          <ArrowDownwardIcon data-testid="arrowDownwardIcon" />
        )}
      </S.IconButton>
    </S.FiltersWrapper>
  );
}
