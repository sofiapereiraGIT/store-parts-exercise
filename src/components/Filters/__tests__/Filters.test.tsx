import React from "react";
import renderer from "react-test-renderer";
import { fireEvent, render } from "@testing-library/react";
import Filters from "../index";

const setSearchFilter = jest.fn();
const setTypeFilter = jest.fn();
const onChangeDescOrder = jest.fn();

describe("Filters Tests", () => {
  it("Renders properly", () => {
    const renderedComponent = renderer.create(
      <Filters
        searchFilter="m"
        setSearchFilter={setSearchFilter}
        loadingTypeOpt={false}
        typeOptions={["Mouse", "Mousepad", "Keyboard", "Monitor"]}
        typeFilter={"Mouse"}
        setTypeFilter={setTypeFilter}
        descOrder={false}
        onChangeDescOrder={onChangeDescOrder}
      />
    );
    expect(renderedComponent.toJSON()).toMatchSnapshot();
  });

  it("Changes search properly ", () => {
    const { getByTestId } = render(
      <Filters
        searchFilter="m"
        setSearchFilter={setSearchFilter}
        loadingTypeOpt={false}
        typeOptions={["Mouse", "Mousepad", "Keyboard", "Monitor"]}
        typeFilter={"Mouse"}
        setTypeFilter={setTypeFilter}
        descOrder={false}
        onChangeDescOrder={onChangeDescOrder}
      />
    );

    const searchField = getByTestId("searchField");
    fireEvent.change(searchField, { target: { value: "mouse 1" } });
    expect(setSearchFilter).toHaveBeenCalledWith("mouse 1");
  });

  it("Changes type filter properly ", () => {
    const { getByRole } = render(
      <Filters
        searchFilter="m"
        setSearchFilter={setSearchFilter}
        loadingTypeOpt={false}
        typeOptions={["Mouse", "Mousepad", "Keyboard", "Monitor"]}
        typeFilter={"Mouse"}
        setTypeFilter={setTypeFilter}
        descOrder={false}
        onChangeDescOrder={onChangeDescOrder}
      />
    );

    const typeDropdown = getByRole("combobox");
    typeDropdown.focus();
    fireEvent.change(document.activeElement as Element, {
      target: { value: "Keyboard" },
    });
    fireEvent.keyDown(document.activeElement as Element, { key: "ArrowDown" });
    fireEvent.keyDown(document.activeElement as Element, { key: "Enter" });
    expect(setTypeFilter).toHaveBeenCalledWith("Keyboard");
  });

  it("Shows descendent order icon properly", () => {
    const { getByTestId } = render(
      <Filters
        searchFilter="m"
        setSearchFilter={setSearchFilter}
        loadingTypeOpt={false}
        typeOptions={["Mouse", "Mousepad", "Keyboard", "Monitor"]}
        typeFilter={"Mouse"}
        setTypeFilter={setTypeFilter}
        descOrder={true}
        onChangeDescOrder={onChangeDescOrder}
      />
    );

    const arrowUpwardIcon = getByTestId("arrowUpwardIcon");
    expect(arrowUpwardIcon).toBeTruthy();
  });

  it("Changes and shows descendent order icon properly", () => {
    const { getByTestId } = render(
      <Filters
        searchFilter="m"
        setSearchFilter={setSearchFilter}
        loadingTypeOpt={false}
        typeOptions={["Mouse", "Mousepad", "Keyboard", "Monitor"]}
        typeFilter={"Mouse"}
        setTypeFilter={setTypeFilter}
        descOrder={false}
        onChangeDescOrder={onChangeDescOrder}
      />
    );

    const arrowDownwardIcon = getByTestId("arrowDownwardIcon");
    expect(arrowDownwardIcon).toBeTruthy();
    const descOrderIcon = getByTestId("descOrderIcon");
    fireEvent.click(descOrderIcon);
    expect(onChangeDescOrder).toHaveBeenCalledTimes(1);
  });
});
