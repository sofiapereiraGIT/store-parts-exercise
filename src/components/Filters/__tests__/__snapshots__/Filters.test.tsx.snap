// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters Tests Renders properly 1`] = `
<div
  className="sc-bczRLJ fIQXYq"
>
  <div
    className="MuiFormControl-root MuiTextField-root sc-gsnTZi kjzJXP"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
      data-shrink={false}
      htmlFor="search"
      id="search-label"
    >
      Search
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputTypeSearch"
        data-testid="searchField"
        disabled={false}
        id="search"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="search"
        value="m"
      />
      <fieldset
        aria-hidden={true}
        className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="PrivateNotchedOutline-legendLabelled-3"
        >
          <span>
            Search
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    aria-owns={null}
    className="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon sc-dkzDqf kkiKvW css-16awh2u-MuiAutocomplete-root"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
  >
    <div
      className="MuiFormControl-root MuiTextField-root sc-gsnTZi kjzJXP MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink={false}
        htmlFor="autocomplete"
        id="autocomplete-label"
      >
        Type
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
        onClick={[Function]}
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-expanded={false}
          aria-invalid={false}
          autoCapitalize="none"
          autoComplete="off"
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
          disabled={false}
          id="autocomplete"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          role="combobox"
          spellCheck="false"
          type="text"
          value=""
        />
        <div
          className="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Open"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            title="Open"
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <span
              className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-3"
          >
            <span>
              Type
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium sc-hKMtZM dKammE css-78trlr-MuiButtonBase-root-MuiIconButton-root"
    data-testid="descOrderIcon"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="arrowDownwardIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="m20 12-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
